propagation/w3c.go

```diff
@@ -0,0 +1,181 @@
+package propagation
+
+import (
+	"context"
+	"time"
+
+	"go.opentelemetry.io/otel/api/kv"
+	"go.opentelemetry.io/otel/api/trace"
+	"google.golang.org/grpc/codes"
+)
+
+// MarshalHoneycombTraceContext uses the information in prop to create trace context headers
+// that conform to the W3C Trace Context specification. The header values are set in headers,
+// which is an HTTPSupplier, an interface to which http.Header is an implementation. The headers
+// are also returned as a map[string]string.
+//
+// Context is passed into this function and returned so that we can maintain the value of the
+// tracestate header. This is required in order to use the Propagator interface exported by the
+// OpenTelemetry Go SDK and avoid writing our own W3C Trace Context parser and serializer.
+//
+// If prop is empty or nil, the return value will be an empty map.
+func MarshalW3CTraceContext(ctx context.Context, prop *PropagationContext) (context.Context, map[string]string) {
+	headerMap := make(map[string]string)
+	otelSpan, err := createOpenTelemetrySpan(prop)
+	if err != nil {
+		return ctx, headerMap
+	}
+	ctx = trace.ContextWithSpan(ctx, otelSpan)
+	propagator := trace.DefaultHTTPPropagator()
+	supplier := newSupplier()
+	propagator.Inject(ctx, supplier)
+	for _, key := range propagator.GetAllKeys() {
+		headerMap[key] = supplier.Get(key)
+	}
+	return ctx, headerMap
+}
+
+// UnmarshalW3CTraceContext parses the information provided in the appropriate headers
+// and creates a PropagationContext instance. Headers are passed in via an HTTPSupplier,
+// which is an interface that defines Get and Set methods, http.Header is an implementation.
+//
+// Context is passed into this function and returned so that we can maintain the value of the
+// tracestate header. This is required in order to use the Propagator interface exported by the
+// OpenTelemetry Go SDK and avoid writing our own W3C Trace Context parser and serializer.
+//
+// If the headers are missing or empty strings, the propagation object will have zero values.
+// In the case of Span IDs and Trace IDs, they will be populated with 16 and 32 character strings
+// containing all zeroes.
+func UnmarshalW3CTraceContext(ctx context.Context, headers map[string]string) (context.Context, *PropagationContext) {
+	supplier := newSupplier()
+	for k, v := range headers {
+		supplier.Set(k, v)
+	}
+	propagator := trace.DefaultHTTPPropagator()
+	ctx = propagator.Extract(ctx, supplier)
```

Comment:
Since this is all within the same package, you have access to the private members of the supplier struct. This block can be changed to avoid allocating a second map and iterating the passed-in map - instead setting the supplier's `values` variable to use the passed in map directly. If you do it that way you can drop the constructor too.

Note that I'm also changing the name here so as to not shadow the type, so the diff is extended to cover where supplier is used.
```suggestion
	supp := supplier{
		values: headers,
	}
	propagator := trace.DefaultHTTPPropagator()
	ctx = propagator.Extract(ctx, supp)
```